plugin/src/main/java/org/opensearch/ml/action/models/TransportUpdateModelAction.java

```diff
@@ -0,0 +1,277 @@
+                    } else {
+                        actionListener
+                            .onFailure(
+                                new MLValidationException(
+                                    "You don't have permission to update the connector, connector id: " + relinkConnectorId
+                                )
+                            );
+                    }
+                }, exception -> {
+                    log.error("Permission denied: Unable to update the connector with ID {}. Details: {}", relinkConnectorId, exception);
+                    actionListener.onFailure(exception);
+                }));
+        } else {
+            actionListener
+                .onFailure(
+                    new IllegalArgumentException("This remote does not have a connector_id field, maybe it uses an internal connector.")
+                );
+        }
+    }
+
+    private void updateRemoteOrTextEmbeddingModel(
+        String modelId,
+        MLUpdateModelInput updateModelInput,
+        MLModel mlModel,
+        User user,
+        UpdateRequest updateRequest,
+        ActionListener<UpdateResponse> actionListener,
+        ThreadContext.StoredContext context
+    ) {
+        String relinkModelGroupId = Strings.hasLength(updateModelInput.getModelGroupId()) ? updateModelInput.getModelGroupId() : null;
+        String relinkConnectorId = Strings.hasLength(updateModelInput.getConnectorId()) ? updateModelInput.getConnectorId() : null;
+
+        if (mlModel.getAlgorithm() == TEXT_EMBEDDING) {
+            if (relinkConnectorId == null) {
+                updateModelWithOrWithoutRelinkModelGroup(modelId, relinkModelGroupId, user, updateRequest, actionListener, context);
+            } else {
+                actionListener
+                    .onFailure(new IllegalArgumentException("Trying to update the connector or connector_id field on a local model"));
+            }
+        } else if (mlModel.getAlgorithm() == REMOTE) {
+            if (relinkConnectorId == null) {
+                updateModelWithOrWithoutRelinkModelGroup(modelId, relinkModelGroupId, user, updateRequest, actionListener, context);
+            } else {
+                updateModelWithRelinkStandAloneConnector(
+                    modelId,
+                    relinkModelGroupId,
+                    relinkConnectorId,
+                    mlModel,
+                    user,
+                    updateRequest,
+                    actionListener,
+                    context
+                );
+            }
+        } else {
+            actionListener
+                .onFailure(
+                    new MLValidationException(
+                        "User doesn't have privilege to perform this operation on this function category: "
+                            + mlModel.getAlgorithm().toString()
+                    )
+                );
+        }
+    }
```

Comment:
What about refactoring to the following? if you early fail as my previous comment mentioned, you can remove the first branch also.

if (mlModel.getAlgorithm() != TEXT_EMBEDDING && mlModel.getAlgorithm() != REMOTE) {
            actionListener
                    .onFailure(
                            new MLValidationException(
                                    "User doesn't have privilege to perform this operation on this function category: "
                                            + mlModel.getAlgorithm().toString()
                            )
                    );
        } else if (relinkConnectorId == null) {
            updateModelWithOrWithoutRelinkModelGroup(modelId, relinkModelGroupId, user, updateRequest, actionListener, context);
        } else if (mlModel.getAlgorithm() == TEXT_EMBEDDING) {
            actionListener
                    .onFailure(new IllegalArgumentException("Trying to update the connector or connector_id field on a local model"));
        } else {
            updateModelWithRelinkStandAloneConnector(
                    modelId,
                    relinkModelGroupId,
                    relinkConnectorId,
                    mlModel,
                    user,
                    updateRequest,
                    actionListener,
                    context
            );
        }

Reply:
I moved it to outer block.